File: android/Ninehoto.Android/Models/AppError.cs
using System;

namespace Ninehoto.Android.Models
{
    public enum AppErrorType
    {
        PhotoLibraryAccessDenied,
        PhotoLibraryAccessRestricted,
        NoPhotosFound,
        ThumbnailLoadFailed,
        DeletionFailed,
        SessionExpired,
        Unknown
    }

    public class AppError : Exception
    {
        public AppErrorType ErrorType { get; }
        public bool IsRecoverable { get; }
        public string? Context { get; }

        public AppError(AppErrorType errorType, string message, bool isRecoverable = true, string? context = null) 
            : base(message)
        {
            ErrorType = errorType;
            IsRecoverable = isRecoverable;
            Context = context;
        }

        public static AppError PhotoLibraryAccessDenied => 
            new AppError(AppErrorType.PhotoLibraryAccessDenied, 
                "Photo library access was denied. Please enable it in Settings.");

        public static AppError PhotoLibraryAccessRestricted => 
            new AppError(AppErrorType.PhotoLibraryAccessRestricted, 
                "Photo library access is restricted on this device.");

        public static AppError NoPhotosFound => 
            new AppError(AppErrorType.NoPhotosFound, 
                "No photos or videos were found in your library.");

        public static AppError ThumbnailLoadFailed(string assetId) => 
            new AppError(AppErrorType.ThumbnailLoadFailed, 
                $"Failed to load thumbnail for asset: {assetId}");

        public static AppError DeletionFailed(int count, string reason) => 
            new AppError(AppErrorType.DeletionFailed, 
                $"Failed to delete {count} item(s): {reason}");

        public static AppError SessionExpired => 
            new AppError(AppErrorType.SessionExpired, 
                "Your session has expired. Please start again.");

        public static AppError FromException(Exception ex) => 
            new AppError(AppErrorType.Unknown, ex.Message, true);
    }
}

File: ios/Ninehoto/Models/AppError.swift
import Foundation

enum AppError: Error, LocalizedError, Equatable {
    case photoLibraryAccessDenied
    case photoLibraryAccessRestricted
    case noPhotosFound
    case thumbnailLoadFailed(assetId: String)
    case deletionFailed(count: Int, reason: String)
    case sessionExpired
    case unknown(underlying: Error)

    var errorDescription: String? {
        switch self {
        case .photoLibraryAccessDenied:
            return "Photo library access was denied. Please enable it in Settings."
        case .photoLibraryAccessRestricted:
            return "Photo library access is restricted on this device."
        case .noPhotosFound:
            return "No photos or videos were found in your library."
        case .thumbnailLoadFailed(let assetId):
            return "Failed to load thumbnail for asset: \(assetId)"
        case .deletionFailed(let count, let reason):
            return "Failed to delete \(count) item(s): \(reason)"
        case .sessionExpired:
            return "Your session has expired. Please start again."
        case .unknown(let underlying):
            return "An unexpected error occurred: \(underlying.localizedDescription)"
        }
    }

    var isRecoverable: Bool {
        switch self {
        case .photoLibraryAccessDenied, .photoLibraryAccessRestricted:
            return true
        case .noPhotosFound:
            return true
        case .thumbnailLoadFailed:
            return true
        case .deletionFailed:
            return true
        case .sessionExpired:
            return true
        case .unknown:
            return true
        }
    }
}